internal/core/operations/switch_active_version/switch_active_version_executor.go

```diff
@@ -0,0 +1,197 @@
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+package switch_active_version
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/topfreegames/maestro/internal/core/entities"
+	"github.com/topfreegames/maestro/internal/core/entities/game_room"
+	"github.com/topfreegames/maestro/internal/core/entities/operation"
+	"github.com/topfreegames/maestro/internal/core/operations"
+	"github.com/topfreegames/maestro/internal/core/services/interfaces"
+	"github.com/topfreegames/maestro/internal/core/services/room_manager"
+	"go.uber.org/zap"
+)
+
+var roomsSlice []*game_room.GameRoom
+
+type SwitchActiveVersionExecutor struct {
+	roomManager        *room_manager.RoomManager
+	schedulerManager   interfaces.SchedulerManager
+	roomsBeingReplaced *sync.Map
+	newCreatedRooms    []*game_room.GameRoom
+}
+
+func NewExecutor(roomManager *room_manager.RoomManager, schedulerManager interfaces.SchedulerManager) *SwitchActiveVersionExecutor {
+	return &SwitchActiveVersionExecutor{
+		roomManager:        roomManager,
+		schedulerManager:   schedulerManager,
+		roomsBeingReplaced: &sync.Map{},
+		newCreatedRooms:    roomsSlice,
+	}
+}
+
+// Execute the process of switching a scheduler active version consists of the following:
+// 1. Creates "replace" goroutines (same number as MaxSurge);
+// 2. Each goroutine will listen to a channel and create a new room using the
+//    new configuration. After the room is ready, it will then delete the room
+//    being replaced;
+// 3. List all game rooms that need to be replaced and produce them into the
+//    replace goroutines channel;
+// 4. Switch the active version
+func (e *SwitchActiveVersionExecutor) Execute(ctx context.Context, op *operation.Operation, definition operations.Definition) error {
+	logger := zap.L().With(
+		zap.String("scheduler_name", op.SchedulerName),
+		zap.String("operation_definition", definition.Name()),
+		zap.String("operation_id", op.ID),
+	)
+	logger.Debug("start switching scheduler active version")
+
+	updateDefinition := definition.(*SwitchActiveVersionDefinition)
```

Comment:
@caiordjesus what do you think about adding a safe check here in this type cast?